frontend/src/app/util/snap_data.cljs

```diff
@@ -0,0 +1,245 @@
+;; This Source Code Form is subject to the terms of the Mozilla Public
+;; License, v. 2.0. If a copy of the MPL was not distributed with this
+;; file, You can obtain one at http://mozilla.org/MPL/2.0/.
+;;
+;; Copyright (c) UXBOX Labs SL
+
+(ns app.util.snap-data
+  "Data structure that holds and retrieves the data to make the snaps. Internaly
+   is implemented with a balanced binary tree that queries by range.
+   https://en.wikipedia.org/wiki/Range_tree"
+  (:require
+   [app.common.data :as d]
+   [app.common.pages :as cp]
+   [app.common.pages.diff :as diff]
+   [app.common.uuid :as uuid]
+   [app.util.geom.grid :as gg]
+   [app.util.geom.snap-points :as snap]
+   [app.util.range-tree :as rt]))
+
+(def snap-attrs [:frame-id :x :y :width :height :hidden :selrect :grids])
+
+;; PRIVATE FUNCTIONS
+
+(defn- make-insert-tree-data
+  [shape-data axis]
+  (fn [tree]
+    (let [tree (or tree (rt/make-tree))]
+      (as-> tree $
+        (reduce (fn [tree data]
+                  (rt/insert tree (get-in data [:pt axis]) data))
+                $ shape-data)))))
+
+(defn- make-delete-tree-data
+  [shape-data axis]
+  (fn [tree]
+    (let [tree (or tree (rt/make-tree))]
+      (as-> tree $
+        (reduce (fn [tree data]
+                  (rt/remove tree (get-in data [:pt axis]) data))
+                $ shape-data)))))
+
+(defn- add-root-frame
+  [page-data]
+  (let [frame-id uuid/zero]
+
+    (-> page-data
+        (assoc-in [frame-id :x] (rt/make-tree))
+        (assoc-in [frame-id :y] (rt/make-tree)))))
+
+(defn- add-frame
+  [page-data frame]
+  (let [frame-id (:id frame)
+        parent-id (:parent-id frame)
+        frame-data (->> (snap/shape-snap-points frame)
+                        (map #(hash-map :type :shape
```

Comment:
forcing use hash-map for maps less than 8 keys is inefficient. My suggestion here is use the `array-map` instead of `hash-map` or let the cljs compiler decide using `#(-> {:type :shape, ...})` or `#(do {:type :shape, ...})`.

Reply:
Also, you using `map` and I think you here want to use `mapv`, because frame-data (on difference with other props) is also associated directly to state (the others are concatenated using concat-vec...)

Reply:
OK.